contracts/V1/StateChannelDiamondProxy/FraudProofFacet.sol

```diff
@@ -47,448 +43,251 @@ contract FraudProofFacet is StateChannelCommon {
     }
 
     // ******************************* FRAUD PROOF IMPLEMENTATION *******************************
-    
-    function _handleBlockDoubleSign(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) internal returns (address) {
-        // Implementation for handling BlockDoubleSign proof
-        return address(0); // Replace with actual logic
+
+
+    // ------------------------------- Block Fraud Proofs --------------------------------------- 
+    function _handleBlockDoubleSign(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) pure internal returns (address) {
+        BlockDoubleSignProof memory blockDoubleSignProof = abi.decode(encodedProof, (BlockDoubleSignProof));
+
+        Block memory block1 = abi.decode(blockDoubleSignProof.block1.encodedBlock, (Block));
+        Block memory block2 = abi.decode(blockDoubleSignProof.block2.encodedBlock, (Block));
+
+        if(fraudProofVerificationContext.channelId != block1.transaction.header.channelId || fraudProofVerificationContext.channelId != block2.transaction.header.channelId) {
+            return address(0);
+        }
+
+        if(block1.stateSnapshotHash != block2.stateSnapshotHash && block1.previousBlockHash != block2.previousBlockHash) {
+            return address(0);
+        }
+        
+        address signer1 = StateChannelUtilLibrary.retriveSignerAddress(
+            blockDoubleSignProof.block1.encodedBlock,
+            blockDoubleSignProof.block1.signature
+        );
+        address signer2 = StateChannelUtilLibrary.retriveSignerAddress(
+            blockDoubleSignProof.block2.encodedBlock,
+            blockDoubleSignProof.block2.signature
+        );
+        if(signer1 != signer2) {
+            return address(0);
+        }
+        return signer1; 
     }
 
-    function _handleBlockEmptyBlock(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) internal returns (address) {
-        // Implementation for handling BlockEmptyBlock proof
-        return address(0); // Replace with actual logic
+    function _handleBlockEmptyBlock(bytes memory encodedProof, FraudProofVerificationContext memory fraudProofVerificationContext) pure internal returns (address) {
+        BlockEmptyProof memory blockEmptyProof = abi.decode(encodedProof, (BlockEmptyProof));
+        Block memory fraudBlock = abi.decode(blockEmptyProof.emptyBlock.encodedBlock, (Block));
+
+        if(fraudProofVerificationContext.channelId != fraudBlock.transaction.header.channelId) {
+            return address(0);
+        }
+        if(fraudBlock.transaction.header.transactionCnt != uint(0)) {
```

Comment:
transactionCnt is blockHeight - It's not the number of transactions in a block. 
An empty block is a one that has 0 transactions or a one that does NOT transition the state, but no error is thrown, otherwise if an error is thrown during the state transition, it would be an invalid state transition. The simplest check that will always hold is if the prior state == posterior state. If this is true -> slash

Compare the snapshots linked to this block and the previous block - in the case that this is the first block (blockHeight == 0) there is no previous block, but the blocks builds on top of the genesisStateSnapshot for the fork and previousBlockHash == genesisStateSnapshotHash

Reply:
ive thought about this, but even on this suggestion check-in  if the prior and posterior state equality is wrong

Reply:
Not sure I understand your comment. Why do you think it's wrong?

The reason why we force non-empty blocks is that the leader election algorithm is a function of the state. If the state doesn't change, the leader is the same. If the leader is the same it's really easy to indefinitely stall by producing empty blocks that don't transition the state.

Reply:
like from your example of simple addition
[1,3,0,0,9]
the block 3,4 will have same state and no transition

Reply:
Yes and we don't want that - so block 4 (addition +0) should be invalid in your example.

Keep in mind that even the simple state machine can use round robin internally, so even if you do +0 and the sum doesn't change, the counter for the leader election will change so state(i) != state(i+1) even though the sum is the same. 

Empty blocks don't transition the state so state(i) == state(i+1) which we don't want.
Even non-empty blocks may not transition the state if the internal state machine logic is implemented incorrectly, but just having round robin internally will always transition the state with non-empty blocks.